include/boost/hana/fwd/index_if.hpp

```diff
@@ -0,0 +1,63 @@
+/*!
+@file
+Forward declares `boost::hana::index_if`.
+
+@copyright Louis Dionne 2013-2017
+@copyright Jason Rice 2017
+Distributed under the Boost Software License, Version 1.0.
+(See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+ */
+
+#ifndef BOOST_HANA_FWD_INDEX_IF_HPP
+#define BOOST_HANA_FWD_INDEX_IF_HPP
+
+#include <boost/hana/config.hpp>
+#include <boost/hana/core/when.hpp>
+
+
+BOOST_HANA_NAMESPACE_BEGIN
+    //! Finds the value associated to the first key satisfying a predicate.
+    //! @ingroup group-Iterable
+    //!
+    //! Given an `Iterable` structure `xs` and a predicate `pred`,
+    //! `index_if(xs, pred)` returns an `IntegralConstant` index *of* the
+    //! first element that satisfies the predicate. If no element satisfies the
+    //! predicate one of two things may occur. For `Foldable` structures the
+    //! result will be an out-of-bounds `IntegralConstant` index, but for
```

Comment:
The documentation here is incorrect. We return an optional index, not an index that is out of bounds. I'm not sure which one is better, actually. I think I'd go with the optional index, and fix the documentation, but I could easily be convinced otherwise.